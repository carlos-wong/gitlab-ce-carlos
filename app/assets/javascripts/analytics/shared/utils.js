import dateFormat from 'dateformat';
import { hideFlash } from '~/flash';
import { slugify } from '~/lib/utils/text_utility';
import { urlQueryToFilter } from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import { dateFormats } from './constants';

export const filterBySearchTerm = (data = [], searchTerm = '', filterByKey = 'name') => {
  if (!searchTerm?.length) return data;
  return data.filter((item) => item[filterByKey].toLowerCase().includes(searchTerm.toLowerCase()));
};

export const toYmd = (date) => dateFormat(date, dateFormats.isoDate);

/**
 * Takes a url and extracts query parameters used for the shared
 * filter bar
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractFilterQueryParameters = (url = '') => {
  /* eslint-disable camelcase */
  const {
    source_branch_name = null,
    target_branch_name = null,
    author_username = null,
    milestone_title = null,
    assignee_username = [],
    label_name = [],
  } = urlQueryToFilter(url);
  /* eslint-enable camelcase */

  return {
    selectedSourceBranch: source_branch_name,
    selectedTargetBranch: target_branch_name,
    selectedAuthor: author_username,
    selectedMilestone: milestone_title,
    selectedAssigneeList: assignee_username,
    selectedLabelList: label_name,
  };
};

/**
 * Takes a url and extracts sorting and pagination query parameters into an object
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractPaginationQueryParameters = (url = '') => {
  const { sort, direction, page } = urlQueryToFilter(url);
  return {
    sort: sort?.value || null,
    direction: direction?.value || null,
    page: page?.value || null,
  };
};

export const getDataZoomOption = ({
  totalItems = 0,
  maxItemsPerPage = 40,
  dataZoom = [{ type: 'slider', bottom: 10, start: 0 }],
}) => {
  if (totalItems <= maxItemsPerPage) {
    return {};
  }

  const intervalEnd = Math.ceil((maxItemsPerPage / totalItems) * 100);

  return dataZoom.map((item) => {
    return {
      ...item,
      end: intervalEnd,
    };
  });
};

export const removeFlash = (type = 'alert') => {
  const flashEl = document.querySelector(`.flash-${type}`);
  if (flashEl) {
    hideFlash(flashEl);
  }
};

/**
 * Prepares metric data to be rendered in the metric_card component
 *
 * @param {MetricData[]} data - The metric data to be rendered
 * @param {Object} popoverContent - Key value pair of data to display in the popover
 * @returns {TransformedMetricData[]} An array of metrics ready to render in the metric_card
 */
export const prepareTimeMetricsData = (data = [], popoverContent = {}) =>
  data.map(({ title: label, identifier, ...rest }) => {
    const metricIdentifier = identifier || slugify(label);
    return {
      ...rest,
      label,
      identifier: metricIdentifier,
      description: popoverContent[metricIdentifier]?.description || '',
    };
  });
