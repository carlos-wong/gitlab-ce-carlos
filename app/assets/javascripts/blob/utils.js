/* global ace */
import Editor from '~/editor/editor_lite';

export function initEditorLite({ el, blobPath, blobContent }) {
  if (!el) {
    throw new Error(`"el" parameter is required to initialize Editor`);
  }
  let editor;

  if (window?.gon?.features?.monacoSnippets) {
    editor = new Editor();
    editor.createInstance({
      el,
      blobPath,
      blobContent,
    });
  } else {
    editor = ace.edit(el);
  }

  return editor;
}

export default () => ({});
