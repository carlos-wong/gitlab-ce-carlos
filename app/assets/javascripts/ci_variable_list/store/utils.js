import { cloneDeep } from 'lodash';
import { displayText, types } from '../constants';

const variableTypeHandler = type =>
  type === displayText.variableText ? types.variableType : types.fileType;

export const prepareDataForDisplay = variables => {
  const variablesToDisplay = [];
  variables.forEach(variable => {
    const variableCopy = variable;
    if (variableCopy.variable_type === types.variableType) {
      variableCopy.variable_type = displayText.variableText;
    } else {
      variableCopy.variable_type = displayText.fileText;
    }
    variableCopy.secret_value = variableCopy.value;

    if (variableCopy.environment_scope === types.allEnvironmentsType) {
      variableCopy.environment_scope = displayText.allEnvironmentsText;
    }
    variablesToDisplay.push(variableCopy);
  });
  return variablesToDisplay;
};

export const prepareDataForApi = (variable, destroy = false) => {
  const variableCopy = cloneDeep(variable);
  variableCopy.protected = variableCopy.protected.toString();
  variableCopy.masked = variableCopy.masked.toString();
  variableCopy.variable_type = variableTypeHandler(variableCopy.variable_type);
  if (variableCopy.environment_scope === displayText.allEnvironmentsText) {
    variableCopy.environment_scope = types.allEnvironmentsType;
  }

  if (destroy) {
    // eslint-disable-next-line
    variableCopy._destroy = destroy;
  }

  return variableCopy;
};

export const prepareEnvironments = environments => environments.map(e => e.name);
