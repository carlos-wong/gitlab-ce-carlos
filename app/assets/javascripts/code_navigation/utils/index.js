export const cachedData = new Map();

export const getCurrentHoverElement = () => cachedData.get('current');
export const setCurrentHoverElement = el => cachedData.set('current', el);

export const addInteractionClass = d => {
  let charCount = 0;
  const line = document.getElementById(`LC${d.start_line + 1}`);
  const el = [...line.childNodes].find(({ textContent }) => {
    if (charCount === d.start_char) return true;
    charCount += textContent.length;
    return false;
  });

  if (el) {
    el.setAttribute('data-char-index', d.start_char);
    el.setAttribute('data-line-index', d.start_line);
    el.classList.add('cursor-pointer', 'code-navigation', 'js-code-navigation');
  }
};
