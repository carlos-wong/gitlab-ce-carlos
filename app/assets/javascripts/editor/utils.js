export const clearDomElement = el => {
  if (!el || !el.firstChild) return;

  while (el.firstChild) {
    el.removeChild(el.firstChild);
  }
};

export default () => ({
  clearDomElement,
});
