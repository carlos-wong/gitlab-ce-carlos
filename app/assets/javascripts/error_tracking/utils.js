/* eslint-disable @gitlab/i18n/no-non-i18n-strings */

/**
 * Tracks snowplow event when user clicks View in Sentry btn
 * @param {String}  externalUrl that will be send as a property for the event
 */
export const trackViewInSentryOptions = url => ({
  category: 'Error Tracking',
  action: 'click_view_in_sentry',
  label: 'External Url',
  property: url,
});

/**
 * Tracks snowplow event when User clicks on error link to Sentry
 * @param {String}  externalUrl that will be send as a property for the event
 */
export const trackClickErrorLinkToSentryOptions = url => ({
  category: 'Error Tracking',
  action: 'click_error_link_to_sentry',
  label: 'Error Link',
  property: url,
});
