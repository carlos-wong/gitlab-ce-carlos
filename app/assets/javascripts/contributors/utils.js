import { getMonthNames } from '~/lib/utils/datetime_utility';

/**
 * Converts provided string to date and returns formatted value as a year for date in January and month name for the rest
 * @param {String}
 * @returns {String}  - formatted value
 *
 * xAxisLabelFormatter('01-12-2019') will return '2019'
 * xAxisLabelFormatter('02-12-2019') will return 'Feb'
 * xAxisLabelFormatter('07-12-2019') will return 'Jul'
 */
export const xAxisLabelFormatter = val => {
  const date = new Date(val);
  const month = date.getUTCMonth();
  const year = date.getUTCFullYear();
  return month === 0 ? `${year}` : getMonthNames(true)[month];
};

/**
 * Formats provided date to YYYY-MM-DD format
 * @param {Date}
 * @returns {String}  - formatted value
 */
export const dateFormatter = date => {
  const year = date.getUTCFullYear();
  const month = date.getUTCMonth();
  const day = date.getUTCDate();

  return `${year}-${`0${month + 1}`.slice(-2)}-${`0${day}`.slice(-2)}`;
};
