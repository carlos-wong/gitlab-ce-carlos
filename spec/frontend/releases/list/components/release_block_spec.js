import { mount } from '@vue/test-utils';
import { first } from 'underscore';
import EvidenceBlock from '~/releases/list/components/evidence_block.vue';
import ReleaseBlock from '~/releases/list/components/release_block.vue';
import ReleaseBlockFooter from '~/releases/list/components/release_block_footer.vue';
import timeagoMixin from '~/vue_shared/mixins/timeago';
import { release } from '../../mock_data';
import Icon from '~/vue_shared/components/icon.vue';
import { scrollToElement } from '~/lib/utils/common_utils';

let mockLocationHash;
jest.mock('~/lib/utils/url_utility', () => ({
  __esModule: true,
  getLocationHash: jest.fn().mockImplementation(() => mockLocationHash),
}));

jest.mock('~/lib/utils/common_utils', () => ({
  __esModule: true,
  scrollToElement: jest.fn(),
}));

describe('Release block', () => {
  let wrapper;
  let releaseClone;

  const factory = (releaseProp, featureFlags = {}) => {
    wrapper = mount(ReleaseBlock, {
      propsData: {
        release: releaseProp,
      },
      provide: {
        glFeatures: {
          releaseIssueSummary: true,
          ...featureFlags,
        },
      },
    });

    return wrapper.vm.$nextTick();
  };

  const milestoneListLabel = () => wrapper.find('.js-milestone-list-label');
  const editButton = () => wrapper.find('.js-edit-button');

  beforeEach(() => {
    releaseClone = JSON.parse(JSON.stringify(release));
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe('with default props', () => {
    beforeEach(() => factory(release));

    it("renders the block with an id equal to the release's tag name", () => {
      expect(wrapper.attributes().id).toBe('v0.3');
    });

    it('renders an edit button that links to the "Edit release" page', () => {
      expect(editButton().exists()).toBe(true);
      expect(editButton().attributes('href')).toBe(release._links.edit_url);
    });

    it('renders release name', () => {
      expect(wrapper.text()).toContain(release.name);
    });

    it('renders release date', () => {
      expect(wrapper.text()).toContain(timeagoMixin.methods.timeFormatted(release.released_at));
    });

    it('renders number of assets provided', () => {
      expect(wrapper.find('.js-assets-count').text()).toContain(release.assets.count);
    });

    it('renders dropdown with the sources', () => {
      expect(wrapper.findAll('.js-sources-dropdown li').length).toEqual(
        release.assets.sources.length,
      );

      expect(wrapper.find('.js-sources-dropdown li a').attributes().href).toEqual(
        first(release.assets.sources).url,
      );

      expect(wrapper.find('.js-sources-dropdown li a').text()).toContain(
        first(release.assets.sources).format,
      );
    });

    it('renders list with the links provided', () => {
      expect(wrapper.findAll('.js-assets-list li').length).toEqual(release.assets.links.length);

      expect(wrapper.find('.js-assets-list li a').attributes().href).toEqual(
        first(release.assets.links).url,
      );

      expect(wrapper.find('.js-assets-list li a').text()).toContain(
        first(release.assets.links).name,
      );
    });

    it('renders author avatar', () => {
      expect(wrapper.find('.user-avatar-link').exists()).toBe(true);
    });

    describe('external label', () => {
      it('renders external label when link is external', () => {
        expect(wrapper.find('.js-assets-list li a').text()).toContain('external source');
      });

      it('does not render external label when link is not external', () => {
        expect(wrapper.find('.js-assets-list li:nth-child(2) a').text()).not.toContain(
          'external source',
        );
      });
    });

    it('renders the footer', () => {
      expect(wrapper.find(ReleaseBlockFooter).exists()).toBe(true);
    });
  });

  it('renders commit sha', () => {
    releaseClone.commit_path = '/commit/example';

    return factory(releaseClone).then(() => {
      expect(wrapper.text()).toContain(release.commit.short_id);

      expect(wrapper.find('a[href="/commit/example"]').exists()).toBe(true);
    });
  });

  it('renders tag name', () => {
    releaseClone.tag_path = '/tag/example';

    return factory(releaseClone).then(() => {
      expect(wrapper.text()).toContain(release.tag_name);

      expect(wrapper.find('a[href="/tag/example"]').exists()).toBe(true);
    });
  });

  it("does not render an edit button if release._links.edit_url isn't a string", () => {
    delete releaseClone._links;

    return factory(releaseClone).then(() => {
      expect(editButton().exists()).toBe(false);
    });
  });

  it('does not render the milestone list if no milestones are associated to the release', () => {
    delete releaseClone.milestones;

    return factory(releaseClone).then(() => {
      expect(milestoneListLabel().exists()).toBe(false);
    });
  });

  it('renders upcoming release badge', () => {
    releaseClone.upcoming_release = true;

    return factory(releaseClone).then(() => {
      expect(wrapper.text()).toContain('Upcoming Release');
    });
  });

  it('slugifies the tag_name before setting it as the elements ID', () => {
    releaseClone.tag_name = 'a dangerous tag name <script>alert("hello")</script>';

    return factory(releaseClone).then(() => {
      expect(wrapper.attributes().id).toBe('a-dangerous-tag-name-script-alert-hello-script');
    });
  });

  describe('evidence block', () => {
    it('renders the evidence block when the evidence is available and the feature flag is true', () =>
      factory(releaseClone, { releaseEvidenceCollection: true }).then(() =>
        expect(wrapper.find(EvidenceBlock).exists()).toBe(true),
      ));

    it('does not render the evidence block when the evidence is available but the feature flag is false', () =>
      factory(releaseClone, { releaseEvidenceCollection: true }).then(() =>
        expect(wrapper.find(EvidenceBlock).exists()).toBe(true),
      ));

    it('does not render the evidence block when there is no evidence', () => {
      releaseClone.evidence_sha = null;

      return factory(releaseClone).then(() => {
        expect(wrapper.find(EvidenceBlock).exists()).toBe(false);
      });
    });
  });

  describe('anchor scrolling', () => {
    beforeEach(() => {
      scrollToElement.mockClear();
    });

    const hasTargetBlueBackground = () => wrapper.classes('bg-line-target-blue');

    it('does not attempt to scroll the page if no anchor tag is included in the URL', () => {
      mockLocationHash = '';
      return factory(release).then(() => {
        expect(scrollToElement).not.toHaveBeenCalled();
      });
    });

    it("does not attempt to scroll the page if the anchor tag doesn't match the release's tag name", () => {
      mockLocationHash = 'v0.4';
      return factory(release).then(() => {
        expect(scrollToElement).not.toHaveBeenCalled();
      });
    });

    it("attempts to scroll itself into view if the anchor tag matches the release's tag name", () => {
      mockLocationHash = release.tag_name;
      return factory(release).then(() => {
        expect(scrollToElement).toHaveBeenCalledTimes(1);

        expect(scrollToElement).toHaveBeenCalledWith(wrapper.element);
      });
    });

    it('renders with a light blue background if it is the target of the anchor', () => {
      mockLocationHash = release.tag_name;

      return factory(release).then(() => {
        expect(hasTargetBlueBackground()).toBe(true);
      });
    });

    it('does not render with a light blue background if it is not the target of the anchor', () => {
      mockLocationHash = '';

      return factory(release).then(() => {
        expect(hasTargetBlueBackground()).toBe(false);
      });
    });
  });

  describe('when the releaseIssueSummary feature flag is disabled', () => {
    describe('with default props', () => {
      beforeEach(() => factory(release, { releaseIssueSummary: false }));

      it('renders the milestone icon', () => {
        expect(
          milestoneListLabel()
            .find(Icon)
            .exists(),
        ).toBe(true);
      });

      it('renders the label as "Milestones" if more than one milestone is passed in', () => {
        expect(
          milestoneListLabel()
            .find('.js-label-text')
            .text(),
        ).toEqual('Milestones');
      });

      it('renders a link to the milestone with a tooltip', () => {
        const milestone = first(release.milestones);
        const milestoneLink = wrapper.find('.js-milestone-link');

        expect(milestoneLink.exists()).toBe(true);

        expect(milestoneLink.text()).toBe(milestone.title);

        expect(milestoneLink.attributes('href')).toBe(milestone.web_url);

        expect(milestoneLink.attributes('title')).toBe(milestone.description);
      });
    });

    it('renders the label as "Milestone" if only a single milestone is passed in', () => {
      releaseClone.milestones = releaseClone.milestones.slice(0, 1);

      return factory(releaseClone, { releaseIssueSummary: false }).then(() => {
        expect(
          milestoneListLabel()
            .find('.js-label-text')
            .text(),
        ).toEqual('Milestone');
      });
    });
  });
});
