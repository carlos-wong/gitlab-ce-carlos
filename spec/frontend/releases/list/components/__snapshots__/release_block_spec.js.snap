// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Release block with default props matches the snapshot 1`] = `
<div
  class="card release-block"
  id="v0.3"
>
  <div
    class="card-body"
  >
    <div
      class="d-flex align-items-start"
    >
      <h2
        class="card-title mt-0 mr-auto"
      >
        
        New release
        
        <!---->
      </h2>
       
      <a
        class="btn btn-default js-edit-button ml-2"
        data-original-title="Edit this release"
        href="http://0.0.0.0:3001/root/release-test/-/releases/v0.3/edit"
        title=""
      >
        <svg
          aria-hidden="true"
          class="s16 ic-pencil"
        >
          <use
            xlink:href="#pencil"
          />
        </svg>
      </a>
    </div>
     
    <div
      class="card-subtitle d-flex flex-wrap text-secondary"
    >
      <div
        class="append-right-8"
      >
        <svg
          aria-hidden="true"
          class="align-middle s16 ic-commit"
        >
          <use
            xlink:href="#commit"
          />
        </svg>
         
        <span
          data-original-title="Initial commit"
          title=""
        >
          c22b0728
        </span>
      </div>
       
      <div
        class="append-right-8"
      >
        <svg
          aria-hidden="true"
          class="align-middle s16 ic-tag"
        >
          <use
            xlink:href="#tag"
          />
        </svg>
         
        <span
          data-original-title="Tag"
          title=""
        >
          v0.3
        </span>
      </div>
       
      <div
        class="js-milestone-list-label"
      >
        <svg
          aria-hidden="true"
          class="align-middle s16 ic-flag"
        >
          <use
            xlink:href="#flag"
          />
        </svg>
         
        <span
          class="js-label-text"
        >
          Milestones
        </span>
      </div>
       
      <a
        class="append-right-4 prepend-left-4 js-milestone-link"
        data-original-title="The 13.6 milestone!"
        href="http://0.0.0.0:3001/root/release-test/-/milestones/2"
        title=""
      >
        
            13.6
          
      </a>
       
            •
          
      <a
        class="append-right-4 prepend-left-4 js-milestone-link"
        data-original-title="The 13.5 milestone!"
        href="http://0.0.0.0:3001/root/release-test/-/milestones/1"
        title=""
      >
        
            13.5
          
      </a>
       
      <!---->
       
      <div
        class="append-right-4"
      >
        
        •
        
        <span
          data-original-title="Aug 26, 2019 5:54pm GMT+0000"
          title=""
        >
          
          released 1 month ago
        
        </span>
      </div>
       
      <div
        class="d-flex"
      >
        
        by
        
        <a
          class="user-avatar-link prepend-left-4"
          href=""
        >
          <span>
            <img
              alt="root's avatar"
              class="avatar s20 "
              data-original-title=""
              data-src="https://www.gravatar.com/avatar/e64c7d89f26bd1972efa854d13d7dd61?s=80&d=identicon"
              height="20"
              src="https://www.gravatar.com/avatar/e64c7d89f26bd1972efa854d13d7dd61?s=80&d=identicon"
              title=""
              width="20"
            />
             
            <div
              aria-hidden="true"
              class="js-user-avatar-image-toolip d-none"
              style="display: none;"
            >
              <div>
                 root 
              </div>
            </div>
          </span>
          <!---->
        </a>
      </div>
    </div>
     
    <div
      class="card-text prepend-top-default"
    >
      <b>
        
        Assets
        
        <span
          class="js-assets-count badge badge-pill"
        >
          5
        </span>
      </b>
       
      <ul
        class="pl-0 mb-0 prepend-top-8 list-unstyled js-assets-list"
      >
        <li
          class="append-bottom-8"
        >
          <a
            class=""
            data-original-title="Download asset"
            href="https://google.com"
            title=""
          >
            <svg
              aria-hidden="true"
              class="align-middle append-right-4 align-text-bottom s16 ic-package"
            >
              <use
                xlink:href="#package"
              />
            </svg>
            
            my link
            
            <span>
              (external source)
            </span>
          </a>
        </li>
        <li
          class="append-bottom-8"
        >
          <a
            class=""
            data-original-title="Download asset"
            href="https://gitlab.com/gitlab-org/gitlab-foss/-/jobs/artifacts/v11.6.0-rc4/download?job=rspec-mysql+41%2F50"
            title=""
          >
            <svg
              aria-hidden="true"
              class="align-middle append-right-4 align-text-bottom s16 ic-package"
            >
              <use
                xlink:href="#package"
              />
            </svg>
            
            my second link
            
            <!---->
          </a>
        </li>
      </ul>
       
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-link"
          data-toggle="dropdown"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="align-top append-right-4 s16 ic-doc-code"
          >
            <use
              xlink:href="#doc-code"
            />
          </svg>
          
          Source code
          
          <svg
            aria-hidden="true"
            class="s16 ic-arrow-down"
          >
            <use
              xlink:href="#arrow-down"
            />
          </svg>
        </button>
         
        <div
          class="js-sources-dropdown dropdown-menu"
        >
          <li>
            <a
              class=""
              href="http://0.0.0.0:3001/root/release-test/-/archive/v0.3/release-test-v0.3.zip"
            >
              Download zip
            </a>
          </li>
          <li>
            <a
              class=""
              href="http://0.0.0.0:3001/root/release-test/-/archive/v0.3/release-test-v0.3.tar.gz"
            >
              Download tar.gz
            </a>
          </li>
          <li>
            <a
              class=""
              href="http://0.0.0.0:3001/root/release-test/-/archive/v0.3/release-test-v0.3.tar.bz2"
            >
              Download tar.bz2
            </a>
          </li>
          <li>
            <a
              class=""
              href="http://0.0.0.0:3001/root/release-test/-/archive/v0.3/release-test-v0.3.tar"
            >
              Download tar
            </a>
          </li>
        </div>
      </div>
    </div>
     
    <div
      class="card-text prepend-top-default"
    >
      <div>
        <p
          data-sourcepos="1:1-1:21"
          dir="auto"
        >
          A super nice release!
        </p>
      </div>
    </div>
  </div>
</div>
`;
